tycho-integration-test/src/stream_processor/rfq_stream_processor.rs

```diff
@@ -0,0 +1,174 @@
+use std::{
+    collections::{HashMap, HashSet},
+    env,
+    fmt::Display,
+    time::Duration,
+};
+
+use miette::{miette, IntoDiagnostic, WrapErr};
+use rand::prelude::IteratorRandom;
+use tokio::{sync::mpsc::Sender, task::JoinHandle};
+use tracing::{info, warn};
+use tycho_common::{
+    models::{token::Token, Chain},
+    Bytes,
+};
+use tycho_simulation::rfq::{
+    protocols::{
+        bebop::{client::BebopClient, client_builder::BebopClientBuilder, state::BebopState},
+        hashflow::{client::HashflowClient, state::HashflowState},
+    },
+    stream::RFQStreamBuilder,
+};
+
+use crate::stream_processor::{StreamUpdate, UpdateType};
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub enum RfqProtocol {
+    Bebop,
+    Hashflow,
+}
+
+impl Display for RfqProtocol {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            RfqProtocol::Bebop => write!(f, "{}", BebopClient::PROTOCOL_SYSTEM),
+            RfqProtocol::Hashflow => write!(f, "{}", HashflowClient::PROTOCOL_SYSTEM),
+        }
+    }
+}
+
+pub struct RfqStreamProcessor {
+    chain: Chain,
+    tvl_threshold: f64,
+    rfq_credentials: HashMap<RfqProtocol, (String, String)>,
+    sample_size: usize,
+    stream_sleep_time: Duration,
+}
+
+impl RfqStreamProcessor {
+    pub fn new(chain: Chain, tvl_threshold: f64, sample_size: usize) -> miette::Result<Self> {
+        let mut rfq_credentials = HashMap::new();
+        let (bebop_user, bebop_key) = (env::var("BEBOP_USER").ok(), env::var("BEBOP_KEY").ok());
+        if let (Some(user), Some(key)) = (bebop_user, bebop_key) {
+            info!("Bebop RFQ credentials found.");
+            rfq_credentials.insert(RfqProtocol::Bebop, (user, key));
+        }
+        let (hashflow_user, hashflow_key) =
+            (env::var("HASHFLOW_USER").ok(), env::var("HASHFLOW_KEY").ok());
+        if let (Some(user), Some(key)) = (hashflow_user, hashflow_key) {
+            info!("Hashflow RFQ credentials found.");
+            rfq_credentials.insert(RfqProtocol::Hashflow, (user, key));
+        }
+        if rfq_credentials.is_empty() {
```

Comment:
this error is only raised if both credentials are missing, if one is there but the other is not this will silently continue

Reply:
That's intentional:
- It will only run the configured protocols, but the binary doesn't require to pass any. Maybe you want to run this binary for regular protocols only.
- If none are provided, then we return an error and we don't start the stream [here](https://github.com/propeller-heads/tycho-simulation/blob/d94ee343b5917020bc3e60fffe8ce384612901a3/tycho-integration-test/src/main.rs#L153)